src/Polar/kernels.jl

```diff
@@ -125,15 +185,11 @@ KA.@kernel function adj_residual_edge_kernel!(
     adj_pinj[fr] -= adj_F[i]
 end
 
-KA.@kernel function adj_residual_node_kernel!(F, adj_F, vm, adj_vm, va, adj_va,
+KA.@kernel function cpu_adj_node_kernel!(F, adj_F, vm, adj_vm, va, adj_va,
```

Comment:
I don't understand why we need two kernels for the CPU and the GPU here?

Reply:
For the CPU kernel both edge metrices are in CSC. For the GPU kernel the transposed edge matrix in COO. You see it in the second loop of the GPU kernel which goes over all non-zeros.

Reply:
I would maybe add a comment to explain that, as this is non-trivial. What do you think?

Reply:
I commented a few functions including these, but then ran out of steam.